extensions/amp-story/0.1/messages.js

```diff
@@ -0,0 +1,187 @@
+  AMP_STORY_SHARING_PROVIDER_NAME_LINKEDIN: 10,
+  AMP_STORY_SHARING_PROVIDER_NAME_PINTEREST: 11,
+  AMP_STORY_SHARING_PROVIDER_NAME_SMS: 12,
+  AMP_STORY_SHARING_PROVIDER_NAME_SYSTEM: 13,
+  AMP_STORY_SHARING_PROVIDER_NAME_TUMBLR: 14,
+  AMP_STORY_SHARING_PROVIDER_NAME_TWITTER: 15,
+  AMP_STORY_SHARING_PROVIDER_NAME_WHATSAPP: 16,
+  AMP_STORY_SYSTEM_LAYER_SHARE_WIDGET_LABEL: 17,
+  AMP_STORY_WARNING_DESKTOP_SIZE_TEXT: 18,
+  AMP_STORY_WARNING_EXPERIMENT_DISABLED_TEXT: 19,
+  AMP_STORY_WARNING_LANDSCAPE_ORIENTATION_TEXT: 20,
+  AMP_STORY_WARNING_UNSUPPORTED_BROWSER_TEXT: 21,
+};
+
+
+/**
+ * @typedef {{
+ *   message: string,
+ *   description: string,
+ * }}
+ */
+let MessageDef;
+
+
+/**
+ * @typedef {!Object<!MessageId, !MessageDef>}
+ */
+export let MessageBundleDef;
+
+
+/**
+ * Language code used if there is no language code specified by the document.
+ * @const {string}
+ */
+const DEFAULT_LANGUAGE_CODE = 'default';
+
+
+/**
+ * @const {!RegExp}
+ */
+const LANGUAGE_CODE_CHUNK_REGEX = /\w+/gi;
+
+
+/**
+ * @param {string} languageCode
+ * @return {!Array<string>} A list of language codes.
+ */
+export function getLanguageCodesFromString(languageCode) {
+  const matches = languageCode.match(LANGUAGE_CODE_CHUNK_REGEX) || [];
+  return matches.reduce((fallbackLanguageCodeList, chunk, index) => {
+    const fallbackLanguageCode = matches.slice(0, index + 1)
+        .join('-')
+        .toLowerCase();
+    fallbackLanguageCodeList.unshift(fallbackLanguageCode);
+    return fallbackLanguageCodeList;
+  }, [DEFAULT_LANGUAGE_CODE]);
+}
+
+
+/**
+ * Gets the message matching the specified message ID in the language specified.
+ * @param {!Object<string, MessageBundleDef>} messageBundles
+ * @param {!Array<string>} languageCodes
+ * @param {!MessageId} messageId
+ */
+function findMessage(messageBundles, languageCodes, messageId) {
+  let message = null;
+
+  languageCodes.some(languageCode => {
+    const messageBundle = messageBundles[languageCode];
+    if (messageBundle && messageBundle[messageId] &&
+        messageBundle[messageId].message) {
+      message = messageBundle[messageId].message;
+      return true;
+    }
+
+    return false;
+  });
+
+  return message;
+}
+
+
+export class MessageService {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    const rootEl = win.document.documentElement;
+
+    /**
+     * @private @const {!Array<string>}
+     */
+    this.rootLanguageCodes_ = this.getLanguageCodesForElement_(rootEl);
+
+    /**
+     * A mapping of language code to message bundle.
+     * @private @const {!Object<string, !MessageBundleDef>}
+     */
+    this.messageBundles_ = {};
+  }
+
+
+  /**
+   * @param {!Element} element
+   * @return {!Array<string>}
+   * @private
+   */
+  getLanguageCodesForElement_(element) {
+    const languageEl = closest(element, el => el.hasAttribute('lang'));
+    const languageCode = languageEl ? languageEl.getAttribute('lang') : null;
```

Comment:
nit: you can probably fallback to empty string here.

Reply:
`getAttribute` can return `null` if the attribute is not present, so I felt `null` was more consistent; languageCode is either the value of the `lang` attribute, or `null` if there wasn't one.

Reply:
sgtm